### PR TITLE
Test with GPU and `m+t4` machines

### PR DESCRIPTION
**Pull request by 0x2b3bfa0**
This pull request replaces the official NVIDIA NGC images (#384, #410) by the ones offered by cloud providers. Now `image = "nvidia"` uses the images described in https://github.com/iterative/terraform-provider-iterative/issues/410#issuecomment-1108447846.

It also removes the fine–grained GPU selectors for Kubernetes, because AKS only provides `accelerator=nvidia` with the node labels, and it's not possible to set that label when creating the cluster.

A solution we can implement, post–release and once tests are green, is weighted [affinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#affinity-and-anti-affinity), prioritizing the exact GPU model but tolerating any NVIDIA GPU device to avoid errors. 

⚠️ **Warning:** testing `k8s` with NVIDIA K80 until Azure lifts the quota restrictions; then, nodes should be moved to `Standard_NC4as_T4_v3`

### CODE REVIEW

**Review by casperdcl — COMMENTED** [review]
Does https://github.com/iterative/terraform-provider-iterative/blob/1ff598265f995c8b11091c38b1699f58d8754703/iterative/resource_runner.go#L377 need to be in `task`/[script.go](https://github.com/iterative/terraform-provider-iterative/blob/master/task/common/machine/script.go)?

**Review by casperdcl — COMMENTED** [review]
can we have docs/wiki for how to list/update these VMs?

related: https://github.com/iterative/terraform-provider-iterative/issues/410#issuecomment-1103767999

File: task/gcp/resources/data_source_image.go [hunk 1]
```diff
@@ -32,7 +32,7 @@ func (i *Image) Read(ctx context.Context) error {
 	image := i.Identifier
 	images := map[string]string{
 		"ubuntu": "ubuntu@ubuntu-os-cloud/ubuntu-2004-lts",
-		"nvidia": "ubuntu@nvidia-ngc-public/nvidia-gpu-cloud-image-20211105",
+		"nvidia": "ubuntu@deeplearning-platform-release/common-cu113-ubuntu-2004",
```
Comment (casperdcl) [inline task/gcp/resources/data_source_image.go:35]: would say cu112 rather than cu113

- mentioned explicitly by https://www.tensorflow.org/install/gpu#software_requirements
- matches AWS DLAMI

https://datascience.stackexchange.com/questions/97849/cuda-for-pytorch-and-cuda-for-tensorflow is for another day :)

Comment (0x2b3bfa0) [reply task/gcp/resources/data_source_image.go:35]: Better if you ask for something [in the realms of possibility](https://cloud.google.com/deep-learning-vm/docs/images); CUDA 11.3 is compatible with TensorFlow 2.8 as per Google Cloud, anyway.

Bonus point for you, because it's not listed on [the TensorFlow compatibility section](https://www.tensorflow.org/install/source#gpu).

**Comment by 0x2b3bfa0** [comment]
> _can we have docs/wiki for how to list/update these VMs?_

See https://github.com/iterative/terraform-provider-iterative/issues/410#issuecomment-1108447846 until further notice.

File: task/common/machine/script.go [hunk 2]
```diff
@@ -102,6 +102,8 @@ fi
 
 rclone copy "$RCLONE_REMOTE/data" /tmp/tpi-task
 
+echo y | /etc/profile.d/install-driver-prompt.sh # for GCP GPU machines
```
Comment (0x2b3bfa0) [inline task/common/machine/script.go:105]:
```suggestion
yes | /etc/profile.d/install-driver-prompt.sh # for GCP GPU machines
```

Yes because... why not?

File: task/gcp/resources/data_source_image.go (hunk 1 above)
Comment (casperdcl) [reply task/gcp/resources/data_source_image.go:35]: ok - in which case probably worth updating the AWS DLAMI to CUDA 11.3 too

Comment (0x2b3bfa0) [reply task/gcp/resources/data_source_image.go:35]: Feel free to, if you have the machine image string.

Comment (0x2b3bfa0) [reply task/gcp/resources/data_source_image.go:35]: Resolved with https://github.com/iterative/terraform-provider-iterative/pull/533/commits/ea2f6ff27b3cb9c6a5f0af6de8a40a976f0ed1d4

File: iterative/resource_task.go [hunk 3]
```diff
@@ -55,7 +55,7 @@ func resourceTask() *schema.Resource {
 				Type:     schema.TypeInt,
 				ForceNew: true,
 				Optional: true,
-				Default:  30,
+				Default:  -1,
```
Comment (casperdcl) [inline iterative/resource_task.go:58]: should this be reverted?

File: task/k8s/resources/resource_job.go [hunk 4]
```diff
@@ -62,13 +62,13 @@ func (j *Job) Create(ctx context.Context) error {
...
-		"m+v100":  "8-64000+nvidia-tesla-v100*1",
-		"l+v100":  "32-256000+nvidia-tesla-v100*4",
-		"xl+v100": "64-512000+nvidia-tesla-v100*8",
+		"m+t4":    "4-16000+nvidia*1",
+		"m+k80":   "4-64000+nvidia*1",
+		"l+k80":   "32-512000+nvidia*8",
+		"xl+k80":  "64-768000+nvidia*16",
+		"m+v100":  "8-64000+nvidia*1",
+		"l+v100":  "32-256000+nvidia*4",
+		"xl+v100": "64-512000+nvidia*8",
```
Comment (casperdcl) [inline task/k8s/resources/resource_job.go:71]: temporary & needs to be reverted?

File: task/k8s/resources/resource_job.go [hunk 5]
```diff
@@ -77,7 +77,7 @@ func (j *Job) Create(ctx context.Context) error {
 	image := j.Attributes.Task.Environment.Image
 	images := map[string]string{
 		"ubuntu": "ubuntu",
-		"nvidia": "nvidia/cuda",
+		"nvidia": "nvidia/cuda:11.3.0-cudnn8-runtime-ubuntu20.04",
```
Comment (casperdcl) [inline task/k8s/resources/resource_job.go:80]:
```suggestion
		"nvidia": "nvidia/cuda:11.3.1-cudnn8-runtime-ubuntu20.04",
```

File: iterative/resource_task.go (hunk 3 above)
Comment (0x2b3bfa0) [reply iterative/resource_task.go:58]: No, -1 means automatic

File: task/k8s/resources/resource_job.go (hunk 4 above)
Comment (0x2b3bfa0) [reply task/k8s/resources/resource_job.go:71]: Unfortunately not temporary, and definitely not to be reverted. 😞 Read the pull request body for details on this tradeoff.

Comment (0x2b3bfa0) [reply task/k8s/resources/resource_job.go:71]: We may want to use `affinity` in a near future, but it's better to make it work out of the box instead of asking users to label their nodes manually.

File: task/aws/resources/resource_launch_template.go [hunk 6]
```diff
@@ -110,6 +109,10 @@ func (l *LaunchTemplate) Create(ctx context.Context) error {
 		},
 	}
 
+	if size := l.Attributes.Size.Storage; size > 0 {
+		input.LaunchTemplateData.BlockDeviceMappings[0].Ebs.VolumeSize = aws.Int32(int32(size))
+	}
```
Comment (DavidGOrtega) [inline task/aws/resources/resource_launch_template.go:114]: Why? Is there any fundamental difference? To avoid 0?

Comment (0x2b3bfa0) [reply task/aws/resources/resource_launch_template.go:114]: Unset means automatic

Comment (DavidGOrtega) [reply task/aws/resources/resource_launch_template.go:114]: Should that end up with different sizes depending on the provider?

Comment (0x2b3bfa0) [reply task/aws/resources/resource_launch_template.go:114]: Yes, providers calculate those automatically from the base image

File: iterative/resource_task.go (hunk 3 above)
Comment (casperdcl) [reply iterative/resource_task.go:58]: would be nice to update the docs too then

- docs/guides/getting-started.md
- docs/resources/task.md
- README.md

all still mention `30` GB and no mention of `-1`

File: task/k8s/resources/resource_job.go (hunk 4 above)
Comment (casperdcl) [reply task/k8s/resources/resource_job.go:71]: ok follow-up issue then

Comment (0x2b3bfa0) [reply task/k8s/resources/resource_job.go:71]: Yes

File: iterative/resource_task.go (hunk 3 above)
Comment (0x2b3bfa0) [reply iterative/resource_task.go:58]: Yes

Comment (casperdcl) [reply iterative/resource_task.go:58]: just pushed a couple of commits - please check

File: docs/resources/task.md [hunk 7]
```diff
@@ -56,7 +56,7 @@ resource "iterative_task" "example" {
 
 - `region` - (Optional) [Cloud region/zone](#cloud-region) to run the task on.
 - `machine` - (Optional) See [Machine Types](#machine-type) below.
-- `disk_size` - (Optional) Size of the ephemeral machine storage in GB.
+- `disk_size` - (Optional) Size of the ephemeral machine storage in GB. `-1`: automatic based on `image`.
```
Comment (casperdcl) [inline docs/resources/task.md:59]: automatic means what? Guessing image but not sure.

File: docs/resources/task.md [hunk 8]
```diff
@@ -169,7 +169,7 @@ In addition to generic types, it's possible to specify any machine type supporte
 The Iterative Provider offers some common machine images which are roughly the same for all supported clouds.
 
 - `ubuntu` - Official [Ubuntu LTS](https://wiki.ubuntu.com/LTS) image (currently 20.04).
-- `nvidia` - Official [NVIDIA NGC](https://docs.nvidia.com/ngc/ngc-deploy-public-cloud)-based images, typically needing `disk_size = 32` GB or more.
+- `nvidia` - Official Ubuntu LTS with CUDA (currently 11.3) images.
```
Comment (casperdcl) [inline docs/resources/task.md:172]:
```suggestion
- `nvidia` - Official Ubuntu LTS with NVIDIA GPU drivers and CUDA toolkit (currently 11.3).
```

File: docs/resources/task.md (hunk 7 above)
Comment (DavidGOrtega) [reply docs/resources/task.md:59]: Agree. More than one might pitfall in not having enough space for the project. 
How does TPI handle not having enough space when syncing data? 🤔

Comment (0x2b3bfa0) [reply docs/resources/task.md:59]: Automatic `disk_size = image_size`

Comment (0x2b3bfa0) [reply docs/resources/task.md:59]: It simplifies considerably the basic use cases, and users can always increase it. Is either that or putting `disk_size = 150` on all the examples.

**Comment by 0x2b3bfa0** [comment]
@iterative/cml, ready for merging? 🚀

File: docs/resources/task.md (hunk 7 above)
Comment (DavidGOrtega) [reply docs/resources/task.md:59]: > It simplifies considerably the basic use cases, and users can always increase it. Is either that or putting disk_size = 150 on all the examples.

I find it better to set a min/default size due to what I say. Users do an example and suddenly it does not work due to sync data

Comment (0x2b3bfa0) [reply docs/resources/task.md:59]: > _I find it better to set a min/default size_

Which would be 150 GB for cross–provider compatibility with the official machine images.

Comment (DavidGOrtega) [reply docs/resources/task.md:59]: Is this change indeed not breaking anything?

Comment (0x2b3bfa0) [reply docs/resources/task.md:59]: Tests are green

Comment (DavidGOrtega) [reply docs/resources/task.md:59]: Speaking about examples / tutorials

Comment (0x2b3bfa0) [reply docs/resources/task.md:59]: > _Suddenly it does not work due to sync data_

It's a risk, of course, but we can't account for the data needs of every user 🙃

Comment (0x2b3bfa0) [reply docs/resources/task.md:59]: > _Speaking about examples / tutorials_

Nope. This change is fully backwards compatible for all the examples/tutorials I am aware of.

Comment (0x2b3bfa0) [reply docs/resources/task.md:59]: In fact, this pull request is the only way for the Jupyter example to work on every cloud

**Review by DavidGOrtega — APPROVED** [review]
👍 

I do not have tested it nor know if the images are OK. Everything looks fine to me so far

File: docs/resources/task.md (hunk 7 above)
Comment (0x2b3bfa0) [reply docs/resources/task.md:59]: Resolved with https://github.com/iterative/terraform-provider-iterative/pull/543#issuecomment-1111031249; free space is not a concern at all.

File: iterative/resource_task.go (hunk 3 above)
Comment (0x2b3bfa0) [reply iterative/resource_task.go:58]: Solved with https://github.com/iterative/terraform-provider-iterative/pull/533/commits/5b96c7e232c4b5c217f8131cc3a4585bfdf67d6c

File: task/k8s/resources/resource_job.go (hunk 4 above)
Comment (0x2b3bfa0) [reply task/k8s/resources/resource_job.go:71]: Moved to https://github.com/iterative/terraform-provider-iterative/issues/158#issuecomment-1111267672
